### PR TITLE
fix loop `inline` bug - cannot read property _walk of null

### LINKED ISSUE #120: TypeError: Cannot read property '_walk' of null
Bug Report

terser v3.9.2

Building Ember js Project

```
Error: Build Canceled: Broccoli Builder ran into an error with `UglifyWriter` plugin. 💥
Cannot read property '_walk' of null
TypeError: Cannot read property '_walk' of null
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1145:28)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1144:24)
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1091:40)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1089:24)
    at AST_Node.eval (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:999:32)
    at Object.TreeWalker._visit (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:1546:21)
    at AST_Node.DEFNODE._walk (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:996:24)
    at walk_body (eval at <anonymous> (/usr/local/src/node_modules/terser/tools/node.js:21:1), <anonymous>:510:17)Pipeline aborted
```

EDIT:
to replicate on a basic ember js project (apparently it was the ember-cli-deploy plugin for me)
```
# npm install ember-cli-deploy
# npm install && ember deploy production
```

### PR DESCRIPTION
**Pull request by fabiosantoscode**
This fixes the "cannot read property _walk of null" error.

### CODE REVIEW

**Comment by kzc** [comment]
Could you please rename the commit and issue to:
```
fix `inline` bug - cannot read property _walk of null

closes #120
```

File: test/compress/issue-120.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
+issue_120: {
+    input: {
```
Comment (kzc) [inline test/compress/issue-120.js]: Test does not fail with unfixed `lib/compress.js`. Please add:
```
    options = {
        defaults: true,
    }
```

File: test/compress/issue-120.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        function f(){for(var t=o=>{var i=+o;return i+i};t(););}
+    }
+    expect_exact: "function f(){for(var t=o=>{var i=+o;return i+i};t(););}"
+}
+
+issue_120_2: {
+    input: {
```
Comment (kzc) [inline test/compress/issue-120.js]: Test does not fail with unfixed `lib/compress.js`. Please add:
```
    options = {
        defaults: true,
    }
```

File: test/compress/issue-120.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        for(var t=o=>{var i=+o;return i+i};t(););
+    }
+    expect_exact: "for(var t=o=>{var i=+o;return i+i};t(););"
+}
```
Comment (kzc) [inline test/compress/issue-120.js]: Please add [this test](https://github.com/terser-js/terser/issues/120#issuecomment-426122195) with `expect_stdout`.

File: lib/compress.js [hunk 4]
```diff
@@ -4926,7 +4926,7 @@ merge(Compressor.prototype, {
             stat = null;
             for (var i = 0; i < len; i++) {
                 var line = body[i];
-                if (line instanceof AST_Var) {
+                if (line instanceof AST_Definitions) {
```
Comment (kzc) [inline lib/compress.js]: Many test failures due to this line. Revert to use `AST_Var`.

File: lib/compress.js [hunk 5]
```diff
@@ -4969,7 +4969,7 @@ merge(Compressor.prototype, {
             var len = fn.body.length;
             for (var i = 0; i < len; i++) {
                 var stat = fn.body[i];
-                if (!(stat instanceof AST_Var)) continue;
+                if (!(stat instanceof AST_Definitions)) continue;
```
Comment (kzc) [inline lib/compress.js]: Many test failures due to this line. Revert to use `AST_Var`.

File: test/compress/issue-120.js (hunk 3 above)
Comment (kzc) [reply test/compress/issue-120.js]:
```
issue_120_3: {
    options = {
        collapse_vars: true,
        evaluate: true,
        inline: 3,
        join_vars: true,
        loops: true,
        reduce_vars: true,
        sequences: true,
        side_effects: true,
        unused: true,
    }
    input: {
        function foo(node) {
            var traverse = function(obj) {
                var i = obj.data;
                return i && i.a != i.b;
            };
            while (traverse(node)) {
                node = node.data;
            }
            return node;
        }
        var x = {
            a: 1,
            b: 2,
            data: {
                a: "hello",
            }
        };
        console.log(foo(x).a, foo({a : "world"}).a);
    }
    expect: {
        function foo(node) {
            for (var traverse = function(obj) {
                var i = obj.data;
                return i && i.a != i.b;
            }; traverse(node); ) node = node.data;
            return node;
        }
        var x = {
            a: 1,
            b: 2,
            data: {
                a: "hello"
            }
        };
        console.log(foo(x).a, foo({
            a: "world"
        }).a);
    }
    expect_stdout: "hello world"
}
```

**Comment by kzc** [comment]
I'm confused. All tests ran cleanly for me locally with https://github.com/terser-js/terser/pull/144/commits/328704ee4c9410c08717e15d1b8d176b4a962c04 without modification.

The changes to past tests https://github.com/terser-js/terser/pull/144/commits/7124178b7a2492ee675b2513cde8a0b669f75df6 is not desirable. If it were needed (which I do not think it is) then it indicates a problem with the fix.

**Comment by kzc** [comment]
I see what's going on now - I took a91fdd20cddb46d87776886cba8c926101f02ed8 and applied the patches up to https://github.com/terser-js/terser/pull/144/commits/328704ee4c9410c08717e15d1b8d176b4a962c04 and ran `npm test` and it *apparently* succeeded. But when I inspected the log there were a lot of `test/compress/*` failures that were not reported - the same failures addressed by https://github.com/terser-js/terser/pull/144/commits/7124178b7a2492ee675b2513cde8a0b669f75df6. So "npm test" in a91fdd20cddb46d87776886cba8c926101f02ed8 was faulty.

With that sorted out I wonder if this inline loop fix can be refined further to limit the changes to code gen.

**Comment by kzc** [comment]
@fabiosantoscode Can you hold off merging this PR for a couple of days? I have some ideas I want to try out to limit when it bails on the `inline` within a loop.

**Comment by fabiosantoscode** [comment]
Sure thing, @kzc

File: lib/compress.js [hunk 6]
```diff
@@ -4941,7 +4941,7 @@ merge(Compressor.prototype, {
             return return_value(stat);
         }
 
-        function can_inject_args(block_scoped, safe_to_inject) {
+        function can_inject_args(block_scoped, defs, safe_to_inject) {
```
Comment (kzc) [inline lib/compress.js]: revert the `can_inject_args` changes - it's a no op.
`defs` is not used.

File: lib/compress.js [hunk 7]
```diff
@@ -5010,8 +5010,8 @@ merge(Compressor.prototype, {
             } while (!(scope instanceof AST_Scope) || scope instanceof AST_Arrow);
             var safe_to_inject = !(scope instanceof AST_Toplevel) || compressor.toplevel.vars;
             var inline = compressor.option("inline");
-            if (!can_inject_vars(block_scoped, inline >= 3 && safe_to_inject)) return false;
-            if (!can_inject_args(block_scoped, inline >= 2 && safe_to_inject)) return false;
+            if (!can_inject_vars(block_scoped, !in_loop && inline >= 3 && safe_to_inject)) return false;
+            if (!can_inject_args(block_scoped, in_loop, inline >= 2 && safe_to_inject)) return false;
```
Comment (kzc) [inline lib/compress.js]: revert the `can_inject_args` changes - it's a no op.

**Comment by kzc** [comment]
If all changes to `lib/compress.js` in this PR are replaced with this patch instead:
```patch
--- a/lib/compress.js
+++ b/lib/compress.js
@@ -4953,2 +4953,3 @@ merge(Compressor.prototype, {
         function can_inject_vars(block_scoped, safe_to_inject) {
+            safe_to_inject = safe_to_inject && (!in_loop || in_loop && !fn.argnames.length);
             var len = fn.body.length;
```
then only 3 tests need be updated instead of the 10 in the latest version of this PR.

Keep in mind that both this patch and the [previous form](https://github.com/terser-js/terser/pull/144/files#diff-230c53e13e3ee367aec7d49470eb42caR5013) are just workarounds for the real problem where a variable symbol definition wasn't cloned correctly (or at all) in the inlined function. But this workaround should suffice until a comprehensive fix is created.

**Comment by kzc** [comment]
Please close this PR in favor of #152.

**Pull request closed without merging**